### PR TITLE
drivers: video: esp32s3: add support for cam interface

### PR DESCRIPTION
**Pull request by LucasTambor**
Add support for the esp32s3 LCD_CAM peripheral.

### CODE REVIEW

**Comment by LucasTambor** [comment]
This commit is a shared work between @epc-ake and me. Mostly based on his PR (#75331) and some internal discussions.

File: drivers/video/video_esp32_dvp.c [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+static struct video_esp32_data esp32_data = {0};
+
+DEVICE_DT_INST_DEFINE(0, &esp32_init, PM_DEVICE_DT_INST_GET(idx), &esp32_data, &esp32_config,
+		      POST_KERNEL, CONFIG_VIDEO_INIT_PRIORITY, &esp32_driver_api);
+
+static int video_esp32_cam_init_clocks(void)
+{
+	int ret = 0;
+
+	ret = pinctrl_apply_state(esp32_config.pcfg, PINCTRL_STATE_DEFAULT);
```
Comment (uLipe) [inline drivers/video/video_esp32_dvp.c:416]: I think since this is not related to clock, it is better to be moved to the global esp32_cam_init?

**Review by uLipe — DISMISSED** [review]
LGTM so far, there is just a small nit, but nothing blocker at all, feel free to skip it.

File: drivers/video/video_esp32_dvp.c (hunk 1 above)
Comment (LucasTambor) [reply drivers/video/video_esp32_dvp.c:416]: Actually I need this guy here to configure the pin that generates the clock for the cam. But it is duplicated, I'll fix it and change the function name.

Comment (marekmatej) [reply drivers/video/video_esp32_dvp.c:410]: Maybe a little comment about this would help future maintenance.

File: drivers/video/video_esp32_dvp.c [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+	cam_ll_enable_invert_de(data->hal.hw, cfg->invert_de);
+	cam_ll_enable_invert_vsync(data->hal.hw, cfg->invert_vsync);
+	cam_ll_enable_invert_hsync(data->hal.hw, cfg->invert_hsync);
+}
+
+static int esp32_init(const struct device *dev)
+{
+	const struct video_esp32_config *cfg = dev->config;
+	struct video_esp32_data *data = dev->data;
+	int ret = 0;
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:348]: `ret is not being used here`.

File: drivers/video/video_esp32_dvp.c [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+	.invert_hsync = DT_INST_PROP(0, invert_hsync),
+	.invert_vsync = DT_INST_PROP(0, invert_vsync),
+	.cam_clk = DT_INST_PROP_OR(0, cam_clk, 0),
+	.clock_dev = DEVICE_DT_GET(DT_INST_CLOCKS_CTLR(0)),
+	.clock_subsys = (clock_control_subsys_t)DT_INST_CLOCKS_CELL(0, offset),
+};
+
+static struct video_esp32_data esp32_data = {0};
+
+DEVICE_DT_INST_DEFINE(0, &esp32_init, PM_DEVICE_DT_INST_GET(idx), &esp32_data, &esp32_config,
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:403]: Would you remove `&` from `&esp32_init` to make it consistent?

File: drivers/video/video_esp32_dvp.c [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+		printk("video pinctrl setup failed (%d)", ret);
+		return ret;
+	}
+
+	/* Enable peripheral */
+	if (!device_is_ready(esp32_config.clock_dev)) {
+		return -ENODEV;
+	}
+
+	clock_control_on(esp32_config.clock_dev, esp32_config.clock_subsys);
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:421]: It is worth checking the result here.

File: drivers/video/video_esp32_dvp.c [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+	if (!esp32_config.cam_clk) {
+		printk("No cam_clk specified\n");
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:424]: I suggest using `LOG_ERR` instead of `printk`.

File: drivers/video/video_esp32_dvp.c [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+	k_fifo_put(&data->fifo_in, vbuf);
+
+	return ret;
+}
+
+static int video_esp32_dequeue(const struct device *dev, enum video_endpoint_id ep,
+			       struct video_buffer **vbuf, k_timeout_t timeout)
+{
+	struct video_esp32_data *data = dev->data;
+	int ret = 0;
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:294]: ' key'  and ' ret' are not used.

File: drivers/video/video_esp32_dvp.c [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+	cam_hal_start_streaming(&data->hal);
+
+	data->is_streaming = true;
+
+	return 0;
+}
+
+static int video_esp32_stream_stop(const struct device *dev)
+{
+	int ret = 0;
```
Comment (sylvioalves) [inline drivers/video/video_esp32_dvp.c:207]: ret are not used.

File: drivers/video/video_esp32_dvp.c (hunk 1 above)
Comment (uLipe) [reply drivers/video/video_esp32_dvp.c:410]: I would prefer self-explanatory code like a function or something, I'm afraid a comment here may get outdated silently until is not reflecting the actual code.

File: drivers/video/video_esp32_dvp.c (hunk 4 above)
Comment (uLipe) [reply drivers/video/video_esp32_dvp.c:421]: I don't think so, enabling or disabling a clock would never fails, enabling a clock that does not exist or is invalid would result in build errors, or am I missing something?

File: drivers/video/video_esp32_dvp.c (hunk 5 above)
Comment (LucasTambor) [reply drivers/video/video_esp32_dvp.c:424]: This function runs on `PRE_KERNEL_2`, logging is not yet available :/

**Review by wmrsouza — DISMISSED** [review]

File: west.yml [hunk 8]
```diff
@@ -157,7 +157,7 @@ manifest:
       groups:
         - hal
     - name: hal_espressif
-      revision: 8c2ae3b0017cadac3de1f57592e879a7bb6ef75d
+      revision: b3072f7c305b6bd789b00db8a0428af0bf18ea6f
```
Comment (sylvioalves) [inline west.yml:160]: Please, update this to latest `e589b872d92d8515ce512b2d34f255d0bccb0e0f` as there is another PR (mcuboot sync) using that revision.

**Review by raffarost — DISMISSED** [review]

**Review by uLipe — DISMISSED** [review]

**Review by loicpoulain — DISMISSED** [review]

**Comment by marekmatej** [comment]
Need to address those comments, otherwise LGTM.

File: drivers/video/video_esp32_dvp.c [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+{
+	const struct video_esp32_config *cfg = dev->config;
+	struct video_esp32_data *data = dev->data;
+
+	k_fifo_init(&data->fifo_in);
+	k_fifo_init(&data->fifo_out);
+	data->config = cfg;
+	video_esp32_cam_ctrl_init(dev);
+
+	if (!device_is_ready(cfg->dma_dev)) {
```
[the remark this answers is not in the document]
Comment (LucasTambor) [reply drivers/video/video_esp32_dvp.c:353]: What do you mean by partial init?

File: dts/xtensa/espressif/esp32s3/esp32s3_common.dtsi [hunk 10]
```diff
@@ -336,6 +336,15 @@
 			status = "disabled";
 		};
 
+		lcd_cam: lcd_cam@60041000 {
+			compatible = "espressif,esp32-lcd-cam";
+			reg = <0x60041000 DT_SIZE_K(4)>;
+			clocks = <&rtc ESP32_LCD_CAM_MODULE>;
+			interrupts = <LCD_CAM_INTR_SOURCE IRQ_DEFAULT_PRIORITY 0>;
```
[the remark this answers is not in the document]
Comment (LucasTambor) [reply dts/xtensa/espressif/esp32s3/esp32s3_common.dtsi:343]: I'd rather leave as is so we got the proper info described here even if we don't use it now on the driver. Later it can be useful when extending the driver.

File: drivers/video/video_esp32_dvp.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (raffarost) [reply drivers/video/video_esp32_dvp.c:349]: it's an error case (not that relevant), but if this condition fails, running the previous inits (e.g. video_esp32_cam_ctrl_init(dev)) could have been avoided beforehand

**Review by raffarost — APPROVED** [review]

**Review by uLipe — APPROVED** [review]

**Review by wmrsouza — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]
